pkg/repo/repo.go

```diff
@@ -127,36 +151,266 @@ func CreateRepoWithMetadata(ctx context.Context, targets []TargetWithMetadata) (
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse cert chain for Fulcio: %w", err)
+		}
+		fulcioAuthorities = append(fulcioAuthorities, *fulcioAuthority)
+	}
+
+	tsaChainPem := concatCertChain(tsaLeaf, tsaIntermed, tsaRoot)
+	tsaAuthorities := []root.CertificateAuthority{}
+	if len(tsaChainPem) > 0 {
+		tsaAuthority, err := certChainToCertificateAuthority(tsaChainPem)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse cert chain for TSA: %w", err)
+		}
+		tsaAuthorities = append(tsaAuthorities, *tsaAuthority)
+	}
+
+	tr, err := root.NewTrustedRoot(
+		root.TrustedRootMediaType01,
+		fulcioAuthorities,
+		ctlogKeys,
+		tsaAuthorities,
+		rekorKeys,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create TrustedRoot: %w", err)
+	}
+	serialized, err := json.Marshal(tr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to serialize TrustedRoot to JSON: %w", err)
+	}
+
+	return &TargetWithMetadata{
+		Name:  "trusted_root.json",
+		Bytes: serialized,
+	}, nil
+}
+
+func pubkeyToTransparencyLogInstance(keyBytes []byte, tm time.Time) (*root.TransparencyLog, string, error) {
+	logID := sha256.Sum256(keyBytes)
+	der, _ := pem.Decode(keyBytes)
+	key, keyDetails, err := getKeyWithDetails(der.Bytes)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return &root.TransparencyLog{
+		BaseURL:             "",
+		ID:                  logID[:],
+		ValidityPeriodStart: tm,
+		HashFunc:            crypto.SHA256, // we can't get this from the keyBytes, assume SHA256
+		PublicKey:           key,
+		SignatureHashFunc:   keyDetails,
+	}, hex.EncodeToString(logID[:]), nil
+}
+
+func getKeyWithDetails(key []byte) (crypto.PublicKey, crypto.Hash, error) {
+	var k any
+	var hashFunc crypto.Hash
+	var err1, err2 error
+
+	k, err1 = x509.ParsePKCS1PublicKey(key)
+	if err1 != nil {
+		k, err2 = x509.ParsePKIXPublicKey(key)
+		if err2 != nil {
+			return 0, 0, fmt.Errorf("can't parse public key with PKCS1 or PKIX: %w, %w", err1, err2)
+		}
+	}
+
+	switch v := k.(type) {
+	case *ecdsa.PublicKey:
+		switch v.Curve {
+		case elliptic.P256():
+			hashFunc = crypto.SHA256
+		case elliptic.P384():
+			hashFunc = crypto.SHA384
+		case elliptic.P521():
+			hashFunc = crypto.SHA512
+		default:
+			return 0, 0, fmt.Errorf("unsupported elliptic curve %T", v.Curve)
+		}
+	case *rsa.PublicKey:
+		switch v.Size() * 8 {
+		case 2048, 3072, 4096:
+			hashFunc = crypto.SHA256
+		default:
+			return 0, 0, fmt.Errorf("unsupported public modulus %d", v.Size())
+		}
+	case *ed25519.PublicKey:
+		hashFunc = crypto.SHA512
+	default:
+		return 0, 0, errors.New("unknown public key type")
+	}
+
+	return k, hashFunc, nil
+}
+
+func certChainToCertificateAuthority(certChainPem []byte) (*root.CertificateAuthority, error) {
+	var cert *x509.Certificate
+	var err error
+	rest := certChainPem
+	certChain := []*x509.Certificate{}
+
+	// skip potential whitespace at end of file (8 is kinda random, but seems to work fine)
+	for len(rest) > 8 {
```

Comment:
Could we use https://pkg.go.dev/strings#TrimSpace instead?

Reply:
Sounds good, I'll do that.